src/include/common/Histogram.h

```diff
@@ -0,0 +1,429 @@
+     */
+    ThreadMinBin(double _accuary, uint64_t _minute_millis) {
+        accuracy = _accuary, minute_millis = _minute_millis;
+    };
+
+    ThreadMinBin(const ThreadMinBin &o) {
+        accuracy = o.accuracy;
+        minute_millis = o.minute_millis;
+        for (auto &tmp : o.per_thread_dist) {
+            per_thread_dist[tmp.first] = tmp.second;
+        }
+    }
+    /* Retrieve the thread-local dist in one given minute. */
+    tdigest::TDigest *GetDistByThreadID(std::thread::id thread_id) {
+        if (per_thread_dist.find(thread_id) == per_thread_dist.end()) {
+            per_thread_dist[thread_id] = tdigest::TDigest(accuracy);
+        }
+        return &per_thread_dist[thread_id];
+    }
+
+    /* Update the value in the distribution of given thread id. */
+    void UpdateDistByThreadID(std::thread::id thread_id, double value) {
+        GetDistByThreadID(thread_id)->add(value);
+    }
+
+    /* Bulk update values in the distribution of given thread id. */
+    void BulkUpdateDistByThreadID(std::thread::id thread_id,
+                                  std::vector<double> means,
+                                  std::vector<double> weights) {
+        if (means.size() != weights.size()) {
+            return;
+        }
+        tdigest::TDigest *cur = GetDistByThreadID(thread_id);
+        for (int i = 0; i < means.size(); i++) {
+            cur->add(means[i], weights[i]);
+        }
+    }
+
+    /* Get list of centroids for dists of all threads in this minute. */
+    std::vector<std::vector<tdigest::Centroid>> GetCentroids() {
+        std::vector<std::vector<tdigest::Centroid>> all_centroids;
+        for (auto &tmp : per_thread_dist) {
+            std::thread::id thread_id = tmp.first;
+            std::vector<tdigest::Centroid> tmp_centroids =
+                per_thread_dist[thread_id].allCentroids();
+            all_centroids.push_back(tmp_centroids);
+        }
+        return all_centroids;
+    }
+
+    /* Convert to Distribution. */
+    std::vector<Distribution> ToDistribution() {
+        std::vector<Distribution> dist;
+        std::vector<std::vector<tdigest::Centroid>> all_centroids =
+            GetCentroids();
+        for (auto &tmp : all_centroids) {
+            Distribution per_thread_dist;
+            per_thread_dist.timestamp = minute_millis;
+            for (auto &c : tmp) {
+                per_thread_dist.centroids.push_back(
+                    std::make_pair(c.mean(), c.weight()));
+            }
+            dist.push_back(per_thread_dist);
+        }
+        return dist;
+    }
+
+    double accuracy;
+    uint64_t minute_millis;
+
+    std::unordered_map<std::thread::id, tdigest::TDigest> per_thread_dist;
+};
+
+/* Wavefront implementation of a histogram. */
+class WavefrontHistogram {
+  public:
+    /**
+     * @description: Construct Wavefront Histogram.
+     * @param cur_min_millis: current miniute in millisecond
+     */
+    WavefrontHistogram(uint64_t cur_min_millis = 0) {
```

Comment:
Should pass a function which provides timestamp instead of just passes a static timestamp